### ISSUE TITLE
"Desktop" menu item not showing up in GNOME 3.38

### ISSUE THREAD

**Issue by teohhanhui** [Rafostar/cast-to-tv-desktop-addon#11]
When `Cast Media` is turned on, the `Desktop` menu item is missing.

`Cast Settings > Modules > Install npm modules` is already done.

**Comment by Rafostar** [comment]
Please share a little more info about your installation.

What is your Linux distro?
What is your version of GNOME Shell?
How did you install the "Desktop Add-on" (from gnome extensions page or from github)?
Do you have the desktop add-on set to enabled in GNOME Extensions App (or in GNOME Tweaks)?

**Comment by teohhanhui** [comment]
> What is your Linux distro?

    $ uname -a
    Linux envy-x360 5.8.16-2-MANJARO #1 SMP PREEMPT Mon Oct 19 11:33:03 UTC 2020 x86_64 GNU/Linux

> What is your version of GNOME Shell?

    $ gnome-shell --version
    GNOME Shell 3.38.1

> How did you install the "Desktop Add-on" (from gnome extensions page or from github)?

Installed from extensions.gnome.org

> Do you have the desktop add-on set to enabled in GNOME Extensions App (or in GNOME Tweaks)?

Yes.

**Comment by Rafostar** [comment]
> GNOME Shell 3.38.1

I have not tested either cast-to-tv extension or any addons on GNOME 3.38 yet. Probably something in 3.38 caused it to be broken. Again. This happens with EVERY new shell release :disappointed:

**Comment by teohhanhui** [comment]
Yeah, that's really unfortunate. Gnome Shell really needs to have some kind of BC promise for extensions.

**Comment by vincentvanvugt** [comment]
@Rafostar Same issue here on GNOME 3.38.1. Some additional info from the logs:
```
JS ERROR: TypeError: AggregateMenu._screencast is undefined
@/home/<user>/.local/share/gnome-shell/extensions/cast-to-tv-desktop-addon@rafostar.github.com/recorder.js:4:19
@/home/<user>/.local/share/gnome-shell/extensions/cast-to-tv-desktop-addon@rafostar.github.com/widget.js:5:33
enableAddon/timeouts[addonName]<@/home/<user>/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/addons.js:59:27
```

This is probably the culprit: https://gitlab.gnome.org/GNOME/gnome-shell/-/commit/2b0731ab81887fd9767da73695c754c5b9aadd7c#ca86d9689a6f7702839217d6fd785f78cea7973a

**Comment by Rafostar** [comment]
@vincentvanvugt 

Thanks. Great find. :+1: 

This confirms that it is indeed GNOME 3.38 related. I would need to port it to DBus to get it back working. So a lot of work to be done to get it back and as usual once this is done it will no longer work on older GNOME versions that did not use DBus for screencast API. Usual GNOME stuff :man_shrugging:.

**Comment by ReubenBTalbott** [comment]
Is there any work around?

**Comment by SwPeter123** [comment]
Is there any update for this? Just bought a Chromecast and realized it does not work with my ubuntu 20.10...
Happy to help if someone can provide a rough guideline on what to do @Rafostar

**Comment by Rafostar** [comment]
@Songyu-Wang 

I already mentioned in my previous comment what needs to be done. Someone has to port this extension code to use DBus connection for remote desktop. I am happy to accept a PR if someone manages, otherwise I will probably work on it myself eventually.

Having to redo your extension from zero every 6 months without prior notice is really frustrating. Trust me :disappointed:

**Comment by devmrfitz** [comment]
Are there any plans to implement this?

**Comment by devmrfitz** [comment]
Also, does Firefox also have to rewrite code everytime GNOME updates so as to share screen on apps like Google Meet?

**Comment by devmrfitz** [comment]
Are there any plans to implement this?
Also, does Firefox also have to rewrite code everytime GNOME updates so as to share screen on websites like Google Meet?
@Rafostar

**Comment by devmrfitz** [comment]
@Rafostar How about PipeWire?

**Comment by Rafostar** [comment]
> @Rafostar How about PipeWire?

Eh. This is not a problem with a GStreamer that is used to do recording underneath. This is a problem with GNOME. Once next version of GNOME Shell is released (the so called GNOME 40) neither this or the Cast to TV extension will work anymore. GNOME developers do not provide any code stability for extensions and they do not care if the shell update will break some extensions.

**Comment by devmrfitz** [comment]
What if this is implemented as a Linux software rather than GNOME extension?

**Comment by bluppfisk** [comment]
Following. Also in gnome 3.38.2

**Comment by maximilize** [comment]
This is the stack trace I get in `journalctl` when trying to activate this addon in current PopOS which is using GNOME 38:

```
Mai 14 22:19:53 pop-os gnome-shell[1524]: JS ERROR: TypeError: can't convert null to object
    @/home/user/.local/share/gnome-shell/extensions/cast-to-tv-desktop-addon@rafostar.github.com/recorder.js:4:39
    @/home/user/.local/share/gnome-shell/extensions/cast-to-tv-desktop-addon@rafostar.github.com/widget.js:7:33
    enableAddon/timeouts[addonName]<@/home/user/.local/share/gnome-shell/extensions/cast-to-tv@rafostar.github.com/addons.js:59:27
```

**Comment by thibaultmol** [comment]
What's the status on pipewire support?

**Comment by hellocatfood** [comment]
There haven't been any updates to this repository in four years, so my guess is no pipewire support any time soon.

**Comment by thibaultmol** [comment]
OH, this was released recently which looks like a good alternative
https://www.phoronix.com/news/GNOME-Network-Displays-0.91
